app/assets/javascripts/angular/controllers/mobile/messages.js

```diff
@@ -0,0 +1,197 @@
+      $scope.resetNewMessage();
+      $scope.messageIsActive = null;
+      $scope.selectedMessage = null;
+    };
+
+    // Fire the request to send a message.
+    $scope.sendMessage = function(message) {
+      // Set us as being the one who sent the message.
+      message.sender = $scope.myCountryId;
+
+      // Error handling
+      $scope.messageError = null;
+      if (message.content === '' || message.content === null) {
+        $scope.messageError = "You must enter a message";
+        return;
+      } else if (message.recipient === null) {
+        $scope.messageError = "You must choose a team";
+        return;
+      }
+
+      // Create the new Message object.
+      $scope.newMessage = { "message": {} }
+      $scope.newMessage.message = angular.copy(message);
+      $scope.newMessage.message.recipient = parseInt(message.recipient);
+
+      $http.post('/api/messages', $scope.newMessage).then(
+      function successCalback(response) {
+        $scope.pushNewMessage(response.data, message.recipient);
+        $scope.resetNewMessage();
+      },
+      function errorCallback(response) {
+
+      });
+    }
+
+    // After submitting a new Message, do some updating to the view.
+    $scope.pushNewMessage = function(savedMessage, conversationPartnerId) {
+      // If there is a conversation already with this partner, grab it.
+      var conversationToUpdate = _.filter($scope.messages, function(message) {
+        return message.conversation_partner.id === conversationPartnerId;
+      });
+
+      if (conversationToUpdate.length > 0) {
+        // Add the new message to the existing messages, update the latest.
+        conversationToUpdate[0].latest_message = savedMessage;
+        conversationToUpdate[0].messages.push(savedMessage);
+      } else {
+        // If this is a new conversation, add the new partner.
+        var conversationPartner = $.grep($scope.teams, function(team) {
+          return team.id == conversationPartnerId;
+        });
+        $scope.messages.push({
+          conversation_partner: conversationPartner[0],
+          latest_message: savedMessage,
+          messages: [savedMessage]
+        });
+      }
+
+      // Sort now that the dates have changed.
+      $scope.messages = _.sortBy($scope.messages, function(conversation) {
+        return conversation.latest_message.updated_at;
+      }).reverse();
+    }
+
+    // Message Filtering
+    $scope.filterFn = function(conversation) {
+      var filter = $scope.messageFilter.id;
+
+      if (filter === 0) {
+        return true;
+      } else if (filter === 1) {
+        return conversation.latest_message.sender_id !== $scope.myCountryId;
+      } else if (filter === 2) {
+        return conversation.latest_message.sender_id === $scope.myCountryId;
+      } else {
+        return true;
+      }
+    };
+
+    $scope.initializeMessages();
+
+    // Group individual messages into Conversations.
+    var groupMessages = function(messages) {
```

Comment:
I brought in Underscore specifically for these changes to messages. It's only 16kb and saved me a nightmare of writing complex and messy code to play with objects and arrays. 

What this function does is put the message data into a way that we want it to be used. It groups messages by who the conversation is with, and adds some more data around the said conversation_partner, as well as the latest_message, and does some sorting on top of that.